src/libraries/System.Private.CoreLib/src/System/Globalization/TextInfo.cs

```diff
@@ -426,93 +426,25 @@ internal static unsafe string ToLowerAsciiInvariant(string s)
                 return string.Empty;
             }
 
-            fixed (char* pSource = s)
+            int i = s.AsSpan().IndexOfAnyInRange('A', 'Z');
+            if (i < 0)
             {
-                int i = 0;
-                while (i < s.Length)
-                {
-                    if (char.IsAsciiLetterUpper(pSource[i]))
-                    {
-                        break;
-                    }
-                    i++;
-                }
-
-                if (i >= s.Length)
-                {
-                    return s;
-                }
-
-                string result = string.FastAllocateString(s.Length);
-                fixed (char* pResult = result)
-                {
-                    for (int j = 0; j < i; j++)
-                    {
-                        pResult[j] = pSource[j];
-                    }
-
-                    pResult[i] = (char)(pSource[i] | 0x20);
-                    i++;
-
-                    while (i < s.Length)
-                    {
-                        pResult[i] = ToLowerAsciiInvariant(pSource[i]);
-                        i++;
-                    }
-                }
-
-                return result;
-            }
-        }
-
-        internal static void ToLowerAsciiInvariant(ReadOnlySpan<char> source, Span<char> destination)
-        {
-            Debug.Assert(destination.Length >= source.Length);
-
-            for (int i = 0; i < source.Length; i++)
-            {
-                destination[i] = ToLowerAsciiInvariant(source[i]);
-            }
-        }
-
-        private static unsafe string ToUpperAsciiInvariant(string s)
-        {
-            if (s.Length == 0)
-            {
-                return string.Empty;
+                return s;
             }
 
             fixed (char* pSource = s)
             {
-                int i = 0;
-                while (i < s.Length)
-                {
-                    if (char.IsAsciiLetterLower(pSource[i]))
-                    {
-                        break;
-                    }
-                    i++;
-                }
-
-                if (i >= s.Length)
-                {
-                    return s;
-                }
-
                 string result = string.FastAllocateString(s.Length);
                 fixed (char* pResult = result)
                 {
-                    for (int j = 0; j < i; j++)
-                    {
-                        pResult[j] = pSource[j];
-                    }
+                    s.AsSpan(0, i).CopyTo(new Span<char>(pResult, result.Length));
 
-                    pResult[i] = (char)(pSource[i] & ~0x20);
+                    pResult[i] = (char)(pSource[i] | 0x20);
                     i++;
 
                     while (i < s.Length)
                     {
-                        pResult[i] = ToUpperAsciiInvariant(pSource[i]);
+                        pResult[i] = ToLowerAsciiInvariant(pSource[i]);
```

Comment:
Do we really want to keep pinning both strings for this operation? Have you tried benchmarking it against `string.Create`?

```cs
return string.Create(s.Length, (s, i), static (result, state) =>
{
    int i = state.i;
    string source = state.s;

    state.s.AsSpan(0, i).CopyTo(result);
    result[i] = (char)(source[i] | 0x20);

    while (i < source.Length)
    {
        result[i] = ToLowerAsciiInvariant(source[i]);
        i++;
    }
});
```

Reply:
Pinning is basically free unless a GC happens to occur while the method is running.  And string.Create, while great for code outside of corelib, requires a delegate, needs to invoke that delegate, etc.